### ISSUE TITLE
fn:uuid

### ISSUE THREAD

**Issue by ChristianGruen** [qt4cg/qtspecs#885]
…to create a random universally unique identifier (UUID), represented as 128-bit value.

Should ideally be nondeterministic, or we may need to do something that’s similar to `fn:random-number-generator`.

**Comment by michaelhkay** [comment]
Perhaps we should add this  as an additional method to fn:random-number-generator, so you can call `fn:random-number-generator()?uuid()`

**Comment by ndw** [comment]
What about other forms of UUID?

If we're only going to support version 4 UUIDs, then I kind of like making it an additional key on `fn:random-number-generator`. My only concern is discoverability.

**Comment by michaelhkay** [comment]
I confess I wasn't aware of the subtleties of different UUID versions and variants. Perhaps someone more familiar with the area could write a requirements statement,

**Comment by ChristianGruen** [comment]
Same for me: I’ve only used type 4 UUIDs.

Adding it to `fn:random-number-generator` would be fine for me. I just wonder what to do with the optional seed? We'd probably ignore it for the UUID?

**Comment by ndw** [comment]
I'm not sure I count as "familiar with the area" exactly, but I've looked into it a few times (for `p:uuid` in XProc, for example).

* Version 1 UUIDs require access to the MAC address of a network card. How practical that is for an implementation may vary.
* Version 2 UUIDs are some DCE security thing. According to the Wikipedia page, they're spec'd in DCE 1.1 Authentication and Security Services which I've never seen
* Version 3 and version 5 are namespace based. These have the nice property that applying the same calculation to the same inputs will produce the same output, but you can't reasonably work out the inputs from the UUID. Cryptographic security on this basis is explicitly disclaimed. (They're only MD5 or SHA-1 hashes, for one thing.)
* Version 4 UUIDs are random

The standard Java UUID class can produce some of these, though it's weirdly underspecified. It can produce version 4 UUIDs, and either version 3 or version 5, but it doesn't specify which, and it could be used to construct version 1 or 2 UUIDs if you worked out the most- and least-significant bits yourself.

**Comment by ChristianGruen** [comment]
@ndw Do you know if these UUID variants are used in practice?

**Comment by ChristianGruen** [comment]
I would tend towards a simple nondeterministic `fn:uuid` function, which could potentially enriched with a version argument if we believe it will be used.

As @line-o pointed out yesterday, `fn:random-number-generator` is ·deterministic·, which makes it cumbersome to generate two different UUIDs. Next, it’s not clear how/if the seed should influence the UUID value.

**Comment by line-o** [comment]
I would also like to bring to ULIDs to the table, which are lexicographically sortable.
Here is an [article comparing the two](https://dev.to/jiisanda/uuid-or-ulid-awesomeness-of-unique-identifiers-48cd).
I am in favor of `fn:uuid()` as `fn:random-number-Generator()` is not a function of time at the moment, which it would become as both UUID v4 and ULID both include the current time stamp.

**Comment by ndw** [comment]
No, @ChristianGruen I have no idea how widely used any flavor of UUID is. My guess is that version 4 dominates. And I'm not going to fuss if we decide to just implement `fn:uuid` that returns version 4 UUIDs. (I do note that the link @line-o points to for ULIDs includes code for version 1, 3, 4, and 5 UUIDs.)

I've never heard of ULIDs before, but I've absolutely wished UUIDs could be sorted, so I'm inclined to favor supporting them. 

But I'm not sure I want `fn:uuid` and `fn:ulid` so we're back to having a function with options, maybe?

**Comment by line-o** [comment]
How about `fn:unique-id($options := ())`

With

- type: UUID, ULID (optional?)
- version: 1 to 5 (applies to UUID only

Default would be UUID v4

**Comment by line-o** [comment]
According to Wikipedia there are 8 versions and two variants of UUIDs. I would propose to make only version 4 mandatory but leave it up to an implementation to allow for all the other ones.
That would make the option argument open-ended as additional information like `namespace` (for v3 and 5) and `variant` would need to be allowed.

**Comment by wendellpiez** [comment]
Cf https://github.com/usnistgov/xslt3-functions/tree/main/random-util (XSLT for v5 UUIDs, with XSpec)

**Comment by ndw** [comment]
At [meeting 107](https://qt4cg.org/meeting/minutes/2025/01-28.html), the CG agreed to close this issue without further action.
